src/vm/jvm/runtime/org/raku/nqp/runtime/LibraryLoader.java

```diff
@@ -40,138 +52,423 @@ public void load(ThreadContext tc, String origFilename) {
+            throw ExceptionHandling.dieInternal(tc, e);
         }
-        catch (Throwable e) {
-            throw ExceptionHandling.dieInternal(tc, e.toString());
+    }
+
+    public static Class<?> loadFile(
+        String fn,
+        ByteClassLoader parent,
+        boolean shared
+    ) throws IOException, IllegalArgumentException, ClassNotFoundException {
+        if (shared)
+            return sharedClasses.computeIfAbsent(fn, (k) -> {
+                try {
+                    return loadFile(k, parent, false);
+                }
+                catch (IOException | IllegalArgumentException | ClassNotFoundException e) {
+                    return null;
+                }
+            });
+        else try (RandomAccessFile fh = new RandomAccessFile(fn, "r")) {
+            switch (fh.readInt()) {
+                case 0xCAFEBABE: // classfile
+                    return loadClass(fn, parent);
+                default: // JAR
+                    // The manifest alone gives us at least an int's width. Let
+                    // the JarInputStream decide if this is a JAR or not.
+                    return loadJar(fn, parent);
+            }
         }
+        catch (EOFException e) {
+            throw new IllegalArgumentException("Unrecognized bytecode format in " + fn);
+        }
+
     }
 
-    private static void loadClass(ThreadContext tc, Class<?> c) throws Throwable {
-        CompilationUnit cu = (CompilationUnit)c.newInstance();
-        cu.shared = tc.gc.sharingHint;
-        cu.initializeCompilationUnit(tc);
-        cu.runLoadIfAvailable(tc);
+    public static Class<?> loadClass(String fn, ByteClassLoader parent) throws IOException, ClassNotFoundException {
+        return new FileClassLoader(fn, parent).loadSerialClass(null);
     }
 
-    private static Class<?> loadJar(byte[] buffer) throws Exception {
-        // This is a (non-empty, non-self-extracting) zip file
-        // These are quite constrained for now
+    public static Class<?> loadJar(String fn, ByteClassLoader parent) throws IOException, ClassNotFoundException {
+        return new JarFileClassLoader(fn, parent).loadSerialClass(null);
+    }
+
+    public static Class<?> loadJar(
+        ByteBuffer buffer,
+        ByteClassLoader parent
+    ) throws IOException, IllegalArgumentException, ClassNotFoundException {
+        String name = null;
+        ByteBuffer classfile = null;
+        ByteBuffer serial = null;
+        try (JarInputStream jis = new JarInputStream(new ByteBufferedInputStream(buffer))) {
+            JarEntry je;
+            while ((je = jis.getNextJarEntry()) != null) {
+                String jf = je.getName();
+                if (jf.endsWith(".class") && classfile == null) {
+                    name = je.getComment();
+                    classfile = readToHeapBuffer(jis);
+                }
+                else if (jf.endsWith(".serialized.lz4") && serial == null)
+                    serial = readToHeapBuffer(jis);
```

Comment:
Likely I'm missing something, but shouldn't this be ```readToHeapBufferLz4(jis)```?

Reply:
It's been a while, but I think the burden of decompressing this is left on whatever depends on `MemoryClassLoader`'s `getResourceAsStream`. This is done to spare memory in its copying, but maybe there's a way to not do that to begin with.

Reply:
I see, so it wasn't an oversight. Thanks for explaining :)